### PR TITLE
Serializing and caching results of AOT compilations at the JITServer

### PR DESCRIPTION
**Pull request by AlexeyKhrabrov**
This PR implements serialization of AOT methods compiled at the JITServer and storing them in its AOT cache.

### CODE REVIEW

**Comment by AlexeyKhrabrov** [comment]
@mpirvu This PR is ready for review.
@dsouzai Could you please review it as well since it touches quite a bit of AOT compiler code?
Thanks.

File: runtime/compiler/control/MethodToBeCompiled.hpp
Comment (mpirvu) [inline runtime/compiler/control/MethodToBeCompiled.hpp:160]: Can we store these fields in `CompilationInfoPerThreadRemote` instead of `TR_MethodToBeCompiled`? The reason is that in some cases we can have thousands of method entries in the queue, but only tens of compilation threads. If, I would prefer a separate commit with this change.

**Review by mpirvu — COMMENTED** [review]
LGTM. I only have some minor comments

**Comment by mpirvu** [comment]
In a follow on PR we may want to add a filter through client command line options, about which methods should be stored in the AOT server side cache: `-Xaot:{methodname}(dontStoreInAOTServerCache)`

**Comment by mpirvu** [comment]
There is a merge conflict due to my latest fix.

**Comment by dsouzai** [comment]
Please don't force push, in the middle of a review right now :)

File: runtime/compiler/codegen/J9AheadOfTimeCompile.cpp
Comment (dsouzai) [inline runtime/compiler/codegen/J9AheadOfTimeCompile.cpp:64]: Could you explain how this is all laid out? `classChainRecord` is of type 
```
class AOTCacheClassChainRecord final : public AOTCacheListRecord<ClassChainSerializationRecord, AOTCacheClassRecord>
```
which means 
```const R *const *records() const { return (const R *const *)_data.end(); }``` 
becomes 
```const AOTCacheClassRecord *const *records() const { return (const AOTCacheClassRecord *const *)_data.end(); }```
where `D _data` is now `ClassChainSerializationRecord _data`. 

Searching for `end()` takes me to `AOTSerializationRecord::end()`, which is
```
const uint8_t *end() const { return (const uint8_t *)this + size(); }
```
I don't really understand how the data is laid out, because at first glance this looks like we're going out of bounds.. I think this needs a bunch of documentation to describe how all of this fits together,

File: runtime/compiler/runtime/RelocationRecord.cpp
Comment (dsouzai) [inline runtime/compiler/runtime/RelocationRecord.cpp:5540]: Might as well fix up the position of the brackets as well. Also maybe add brackets for the `if` clause.

File: runtime/compiler/control/MethodToBeCompiled.hpp
Comment (AlexeyKhrabrov) [reply runtime/compiler/control/MethodToBeCompiled.hpp:160]: Yes, I can move these to `compInfoPT`. If memory footprint of this structure is a concern, its size can be reduced by up to 16 bytes by storing boolean flags as bits and reordering some fields (maybe in a future PR).

**Comment by AlexeyKhrabrov** [comment]
> In a follow on PR we may want to add a filter through client command line options, about which methods should be stored in the AOT server side cache: `-Xaot:{methodname}(dontStoreInAOTServerCache)`

Yes, that would be helpful for diagnostics etc., I'll add the option in a future PR. Should there be separate options for store and load or just a single one for bypassing AOT cache altogether for a given method?

File: runtime/compiler/control/JITClientCompilationThread.cpp
Comment (dsouzai) [inline runtime/compiler/control/JITClientCompilationThread.cpp:3151]: Why is the `-1` needed?

File: runtime/compiler/control/JITClientCompilationThread.cpp
Comment (dsouzai) [inline runtime/compiler/control/JITClientCompilationThread.cpp:3197]: Worth noting that this isn't going to be an index but a pointer diff. I think you have to divide by the size of the array element to get an index. Does the rest of the code assume it's an array index or a pointer diff?

**Comment by AlexeyKhrabrov** [comment]
> There is a merge conflict due to my latest fix.

I guess we also need to wait until #13973 is merged because JITServer tests would currently fail.

**Comment by mpirvu** [comment]
Yes, I will not run tests before #13973 is merged and this code rebased.

**Review by dsouzai — CHANGES_REQUESTED** [review]
Overall looks reasonable; however I have a few questions/suggestions. Also, you should add doxygen comments for all the new APIs  you added (including the helper methods).

File: runtime/compiler/arm/codegen/J9AheadOfTimeCompile.cpp
Comment (AlexeyKhrabrov) [reply runtime/compiler/arm/codegen/J9AheadOfTimeCompile.cpp:247]: AOT cache identifies class loaders by name of the 1st loaded class, not by its class chain. This class loader ID is obtained from a class chain record as the class loader ID of the root class of the chain. The setter implementations in relocation records do the right thing and call `addClassChainSerializationRecord()` for class chains and `addClassLoaderSerializationRecord()` for class chains identifying loaders. I'll explain this in the doc comments that I'll add for the `add*SerializationRecord()` methods.

File: runtime/compiler/codegen/J9AheadOfTimeCompile.cpp
Comment (AlexeyKhrabrov) [reply runtime/compiler/codegen/J9AheadOfTimeCompile.cpp:64]: The layout of `AOTCacheClassChainRecord` is as follows:
- `AOTCacheClassChainRecord` header
- `ClassChainSerializationRecord` header
- `ClassSerializationRecord` data: array of N class IDs
- array of N `AOTCacheClassRecord` pointers

I'll add comments in `JITServerAOTCache.hpp` to better describe the layout. I'll also move `classChainRecord->records()[0]` to a separate helper method in `AOTCacheClassChainRecord`.

File: runtime/compiler/control/JITClientCompilationThread.cpp
Comment (AlexeyKhrabrov) [reply runtime/compiler/control/JITClientCompilationThread.cpp:3151]: `classChain[0]` stores the size of the whole class chain record in bytes, including the 1st word itself. `len` is the number of classes in the chain.

File: runtime/compiler/control/JITClientCompilationThread.cpp
Comment (AlexeyKhrabrov) [reply runtime/compiler/control/JITClientCompilationThread.cpp:3197]: It is actually the index, and the rest of the code assumes that it's the index. I'm pretty sure this is standard C and C++, and it's true at least with gcc/g++. But I guess I can rewrite this as diff between pointers converted to `uint8_t *` divided by `sizeof` to make it less ambiguous.

Comment (dsouzai) [reply runtime/compiler/control/JITClientCompilationThread.cpp:3197]: No I think if it's standard C/C++ then it's cleaner code as it stands.

**Comment by mpirvu** [comment]
This PR is ready for a rebase because #13793 was merged.

**Comment by AlexeyKhrabrov** [comment]
Addressed the comments and rebased.

**Comment by mpirvu** [comment]
Jenkins compile all jdk8

File: runtime/compiler/control/MethodToBeCompiled.hpp
Comment (mpirvu) [inline runtime/compiler/control/MethodToBeCompiled.hpp:48]: I think this declaration can be removed now

Comment (AlexeyKhrabrov) [reply runtime/compiler/control/MethodToBeCompiled.hpp:48]: True. Should I push this change now or wait until the tests are done?

**Comment by mpirvu** [comment]
Let's wait for the compilation tests. Frankly, I thought they would be done by now.

**Comment by mpirvu** [comment]
aarch64_linux compilation is failing with
```
09:00:38  /home/jenkins/workspace/Build_JDK8_aarch64_linux_Personal/openj9/runtime/compiler/aarch64/codegen/J9AheadOfTimeCompile.cpp: In member function 'virtual void J9::ARM64::AheadOfTimeCompile::processRelocations()':
09:00:38  /home/jenkins/workspace/Build_JDK8_aarch64_linux_Personal/openj9/runtime/compiler/aarch64/codegen/J9AheadOfTimeCompile.cpp:75:45: error: invalid use of member function 'TR::Compilation* OMR::AheadOfTimeCompile::comp()' (did you forget the '()' ?)
09:00:38            TR::SymbolValidationManager *svm = comp->getSymbolValidationManager();
09:00:38                                               ^~~~
09:00:38  /home/jenkins/workspace/Build_JDK8_aarch64_linux_Personal/openj9/runtime/compiler/aarch64/codegen/J9AheadOfTimeCompile.cpp:75:49: error: base operand of '->' is not a pointer
09:00:38            TR::SymbolValidationManager *svm = comp->getSymbolValidationManager();
09:00:38                                                   ^~
09:00:38  runtime/compiler/CMakeFiles/j9jit.dir/build.make:2476: recipe for target 'runtime/compiler/CMakeFiles/j9jit.dir/aarch64/codegen/J9AheadOfTimeCompile.cpp.o' failed
09:00:38  gmake[4]: *** [runtime/compiler/CMakeFiles/j9jit.dir/aarch64/codegen/J9AheadOfTimeCompile.cpp.o] Error 1
09:00:38  gmake[4]: *** Waiting for unfinished jobs....
```

**Comment by AlexeyKhrabrov** [comment]
> `comp` needs to be replaced with `self()->comp()`

Oops, copy-paste mistake during refactoring. Will push the fix shortly.

**Comment by AlexeyKhrabrov** [comment]
Done. By the way, most of the code in `AheadOfTimeCompile::processRelocations()` is the same across architectures and can be factored out.

**Comment by mpirvu** [comment]
Jenkins compile all jdk8

**Comment by mpirvu** [comment]
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

**Review by mpirvu — APPROVED** [review]

**Comment by mpirvu** [comment]
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

**Comment by mpirvu** [comment]
Tests on zLinux passed. 
Tests on pLinux failed due to infra
Tests on xLinux timedout 
```
13:26:30  Found a total of 20 nodes with the 'ci.role.test&&hw.arch.x86&&sw.os.linux' label
[Pipeline] echo
13:26:30  machine limit is 20
[Pipeline] timeout
13:26:30  Timeout set to expire in 1 hr 0 min
[Pipeline] {
[Pipeline] copyArtifacts
[Pipeline] }
[Pipeline] // timeout
[Pipeline] echo
13:26:31  Cannot get cached TRSS JSON data. Skipping copyArtifacts...
[Pipeline] timeout
13:26:31  Timeout set to expire in 2 hr 0 min
[Pipeline] {
[Pipeline] copyArtifacts
15:26:31  Cancelling nested steps due to timeout
[Pipeline] }
[Pipeline] // timeout
[Pipeline] echo
15:26:31  Cannot run copyArtifacts from test.getDependency. Skipping copyArtifacts...
[Pipeline] sh
[Pipeline] }
[Pipeline] // timestamps
[Pipeline] }
[Pipeline] // stage
[Pipeline] echo
terminateJavaProcesses iteration: 1
[Pipeline] echo
PROCESSCATCH: Checking for any leftover java processes on the machine
[Pipeline] sh
[Pipeline] }
[Pipeline] // timeout
[Pipeline] echo
Exception: java.lang.InterruptedException
[Pipeline] }
[Pipeline] // node
[Pipeline] }
[Pipeline] // stage
[Pipeline] End of Pipeline
Finished: FAILURE
```

**Comment by mpirvu** [comment]
Jenkins compile all jdk8

**Comment by mpirvu** [comment]
Jenkins compile all jdk8

**Comment by mpirvu** [comment]
All compilation tests have passed as well

**Comment by mpirvu** [comment]
@dsouzai If you agree, please approve this PR. Thanks

**Review by dsouzai — APPROVED** [review]

**Pull request merged**
